recipes/Voicebank/MTL/CoopNet/CoopNet.py

```diff
@@ -0,0 +1,224 @@
+"""Classes and utils needed for the cooperative-learning recipe
+"""
+import torch
+import logging
+from collections import OrderedDict
+from speechbrain.utils import checkpoints
+
+from torch import nn
+from speechbrain.processing.features import STFT
+from speechbrain.lobes.models.transformer.Transformer import get_lookahead_mask
+from speechbrain.lobes.models.transformer.TransformerSE import CNNTransformerSE
+from speechbrain.lobes.features import Fbank
+
+logger = logging.getLogger(__name__)
+
+
+def torch_save(obj, path):
+    """Saves the obj's parameters to path.
+
+    Default save hook for torch.nn.Modules
+    For saving torch.nn.Module state_dicts.
+
+    Arguments
+    ---------
+    obj : torch.nn.Module
+        Instance to save.
+    path : str, pathlib.Path
+        Path where to save to.
+
+    Returns
+    -------
+    None
+        State dict is written to disk.
+    """
+    state_dict = obj.state_dict()
+    if not state_dict:
+        logger.warning(f"Saving an empty state_dict for {obj} in {path}.")
+    torch.save(state_dict, path)
+
+
+def torch_recovery(obj, path, device=None):
+    """Loads a torch.nn.Module state_dict from the given path instantly.
+
+    This can be made the default for torch.nn.Modules with:
+    >>> DEFAULT_LOAD_HOOKS[torch.nn.Module] = torch_recovery
+
+    Arguments
+    ---------
+    obj : torch.nn.Module
+        Instance for which to load the parameters.
+    path : str, pathlib.Path
+        Path where to load from.
+    end_of_epoch : bool
+        Whether the recovery comes from an end of epoch checkpoint.
+    device : str
+        Torch device, where to map the loaded parameters.
+
+    Returns
+    -------
+    None
+        Given object is modified in place.
+    """
+    try:
+        obj.load_state_dict(torch.load(path, map_location=device), strict=True)
+    except TypeError:
+        obj.load_state_dict(torch.load(path, map_location=device))
+
+
+@checkpoints.register_checkpoint_hooks
+class MultipleOptimizer(object):
+    """Class that allows the initialization of multiple optimization methods, and handles gradient updates."""
+
+    def __init__(self, params, init_keys=["se_opt", "asr_opt"], **optim):
+        self.optimizers = OrderedDict(
+            {key: optim[key](params[key]) for key in init_keys}
+        )
+
+    def zero_grad(self):
+        for op in self.optimizers.values():
+            op.zero_grad()
+
+    def step(self):
+        for op in self.optimizers.values():
+            op.step()
+
+    @checkpoints.mark_as_saver
+    def save(self, path):
+        key = path.stem.split("optimizer_")[1]
+        torch_save(self.optimizers[key], path)
+
+    @checkpoints.mark_as_loader
+    def load(self, path, end_of_epoch, device):
+        del end_of_epoch
+        del device
+        key = path.stem.split("optimizer_")[1]
+        torch_recovery(self.optimizers[key], path)
```

Comment:
Just a quick question: This looks like the multioptimizer needs to be added multiple times to the checkpointer. Couldn't this class just save all its optimizers in one file?

Reply:
Hi Gastron, 
Sorry for the late reply.
From what I remember I set this up in this way because I thought that it could be useful to retrieve an optimizer directly from a file instead of loading all the optimizers and selecting one. 
I did not end up using it, so I could change it if a change is desirable.